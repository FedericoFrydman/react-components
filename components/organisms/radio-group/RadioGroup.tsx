import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './radio-group.scss';
import { Heading } from '../../atoms/heading/Heading';
import { Icon } from '../../atoms/icon/Icon';
import { Text } from '../../../components/atoms/text/Text';

interface Option {
  name: string;
  price: number;
  value: string;
}

export interface RadioGroupProps extends Omit<React.HTMLProps<HTMLDivElement>, 'onClick' | 'css'> {
  /**
   * An event to handle option selection.
   */
  onClick?: (option: Option) => void;
  /**
   * A list of options for RadioGroup.
   */
  options?: Option[];
  /**
   * title props to show label.
   */
  title?: string;
  /**
   * Value props to show selected option
   */
  value?: string;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const RadioGroupComponent = (
  { title, options, value, onClick, classNames = [], ...props }: RadioGroupProps,
  ref: RadioGroupProps['ref'],
) => {
  const classCheckbox = 'lsux-checkbox';
  const classOption = 'lsux-option';

  return (
    <div {...props} className={cn('lsux-rbg-container', 'checkbox-section', classNames)} ref={ref}>
      {title && <Heading as="T20" className="lsux-checkbox-title" text={title} />}
      <div className={`${classCheckbox}-options-section`}>
        <ul className={`${classCheckbox}-option-div`}>
          {options &&
            options?.length &&
            options.map((option, index) => (
              <li className={`${classOption}-li`} key={option.name} onClick={() => onClick && onClick(option)}>
                <div className={`${classOption}-div`}>
                  <div className={`${classOption}-subdiv`}>
                    <label htmlFor={'helper-radio-' + index}>
                      <Text textWeight="normal" text={option.name} classNames={['mb-1']} />
                      <Text textWeight="semibold" text={`$${option.price}/mo`} />
                    </label>
                  </div>
                  <div className="lsux-checked-icon">
                    {value === option.value ? (
                      <Icon name="warning_circle_check" color={'P200'} size="medium" />
                    ) : (
                      <Icon name="warning_circle" color={'N700'} size="medium" />
                    )}
                  </div>
                </div>
              </li>
            ))}
        </ul>
      </div>
    </div>
  );
};

/**
 * A radio button is a quick way to extract a single coded answer from our users by presenting multiple possibilities and allowing only one option to be chosen.
 */

export const RadioGroup = forwardRef<HTMLDivElement, RadioGroupProps>(RadioGroupComponent);
