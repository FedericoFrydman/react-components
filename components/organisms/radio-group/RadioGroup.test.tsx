import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { shallow, ShallowWrapper } from 'enzyme';
import { RadioGroup, RadioGroupProps } from './RadioGroup';

describe('RadioGroup Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test-rbg';
    wrapper = shallow(
      <RadioGroup
        onClick={() => alert('Radio button was pressed')}
        options={[
          {
            name: 'Trial Defense',
            price: 14.95,
            value: 'Trial Defense',
          },

          { name: 'Gun Owner', price: 14.95, value: 'Gun Owner' },
          { name: 'Ride Share', price: 14.95, value: 'Ride Share' },
          {
            name: 'Home Business',
            price: 14.95,
            value: 'Home Business',
          },
          {
            name: 'Business Builder',
            price: 14.95,
            value: 'Business Builder',
          },
        ]}
        title="Plan add-ons"
        value="Home Business"
      />,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the RadioGroup the provided classNames', () => {
          const rbgElem = renderRadioGroup({ classNames });

          expect(rbgElem).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the snackbar', () => {
      expect(renderRadioGroup({ ref })).toEqual(ref.current);
    });
  });

  const renderRadioGroup = (props: RadioGroupProps = {}): HTMLElement => {
    render(<RadioGroup {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
