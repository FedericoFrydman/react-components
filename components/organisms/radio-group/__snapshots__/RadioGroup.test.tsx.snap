// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup Component should render the component 1`] = `
<div
  className="lsux-rbg-container checkbox-section"
>
  <ForwardRef(HeadingComponent)
    as="T20"
    className="lsux-checkbox-title"
    text="Plan add-ons"
  />
  <div
    className="lsux-checkbox-options-section"
  >
    <ul
      className="lsux-checkbox-option-div"
    >
      <li
        className="lsux-option-li"
        key="Trial Defense"
        onClick={[Function]}
      >
        <div
          className="lsux-option-div"
        >
          <div
            className="lsux-option-subdiv"
          >
            <label
              htmlFor="helper-radio-0"
            >
              <ForwardRef(TextComponent)
                classNames={
                  Array [
                    "mb-1",
                  ]
                }
                text="Trial Defense"
                textWeight="normal"
              />
              <ForwardRef(TextComponent)
                text="$14.95/mo"
                textWeight="semibold"
              />
            </label>
          </div>
          <div
            className="lsux-checked-icon"
          >
            <ForwardRef(IconComponent)
              color="N700"
              name="warning_circle"
              size="medium"
            />
          </div>
        </div>
      </li>
      <li
        className="lsux-option-li"
        key="Gun Owner"
        onClick={[Function]}
      >
        <div
          className="lsux-option-div"
        >
          <div
            className="lsux-option-subdiv"
          >
            <label
              htmlFor="helper-radio-1"
            >
              <ForwardRef(TextComponent)
                classNames={
                  Array [
                    "mb-1",
                  ]
                }
                text="Gun Owner"
                textWeight="normal"
              />
              <ForwardRef(TextComponent)
                text="$14.95/mo"
                textWeight="semibold"
              />
            </label>
          </div>
          <div
            className="lsux-checked-icon"
          >
            <ForwardRef(IconComponent)
              color="N700"
              name="warning_circle"
              size="medium"
            />
          </div>
        </div>
      </li>
      <li
        className="lsux-option-li"
        key="Ride Share"
        onClick={[Function]}
      >
        <div
          className="lsux-option-div"
        >
          <div
            className="lsux-option-subdiv"
          >
            <label
              htmlFor="helper-radio-2"
            >
              <ForwardRef(TextComponent)
                classNames={
                  Array [
                    "mb-1",
                  ]
                }
                text="Ride Share"
                textWeight="normal"
              />
              <ForwardRef(TextComponent)
                text="$14.95/mo"
                textWeight="semibold"
              />
            </label>
          </div>
          <div
            className="lsux-checked-icon"
          >
            <ForwardRef(IconComponent)
              color="N700"
              name="warning_circle"
              size="medium"
            />
          </div>
        </div>
      </li>
      <li
        className="lsux-option-li"
        key="Home Business"
        onClick={[Function]}
      >
        <div
          className="lsux-option-div"
        >
          <div
            className="lsux-option-subdiv"
          >
            <label
              htmlFor="helper-radio-3"
            >
              <ForwardRef(TextComponent)
                classNames={
                  Array [
                    "mb-1",
                  ]
                }
                text="Home Business"
                textWeight="normal"
              />
              <ForwardRef(TextComponent)
                text="$14.95/mo"
                textWeight="semibold"
              />
            </label>
          </div>
          <div
            className="lsux-checked-icon"
          >
            <ForwardRef(IconComponent)
              color="P200"
              name="warning_circle_check"
              size="medium"
            />
          </div>
        </div>
      </li>
      <li
        className="lsux-option-li"
        key="Business Builder"
        onClick={[Function]}
      >
        <div
          className="lsux-option-div"
        >
          <div
            className="lsux-option-subdiv"
          >
            <label
              htmlFor="helper-radio-4"
            >
              <ForwardRef(TextComponent)
                classNames={
                  Array [
                    "mb-1",
                  ]
                }
                text="Business Builder"
                textWeight="normal"
              />
              <ForwardRef(TextComponent)
                text="$14.95/mo"
                textWeight="semibold"
              />
            </label>
          </div>
          <div
            className="lsux-checked-icon"
          >
            <ForwardRef(IconComponent)
              color="N700"
              name="warning_circle"
              size="medium"
            />
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
